### ISSUE TITLE
Replace special chars with underscores for environment variables with credentials

### ISSUE THREAD

**Issue by bryant-finney** [python-poetry/poetry#7123]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) stable Poetry version, installed using a recommended method.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have searched the [FAQ](https://python-poetry.org/docs/faq/) and general [documentation](https://python-poetry.org/docs/) and believe that my question is not already covered.

- **Poetry version**: [>=1.0.0](https://github.com/python-poetry/poetry/releases/tag/1.0.0)
- **Python version**: `3`
- **OS version and name**: N/A
- **`pyproject.toml`**: [bryant-finney/pyproject.toml](https://gist.github.com/bryant-finney/c2537462fea99f1217a7ab113c56edd6)

## Request for Enhancement

Currently, I am unable to use environment variables to authenticate for installing packages from a private internal PyPI repository `example.com`.

In the linked [`pyproject.toml`](https://gist.github.com/bryant-finney/c2537462fea99f1217a7ab113c56edd6), the source repository named `example.com` provides a private package named `the-package`:

```toml
# ...

[[tool.poetry.source]]
name = "example.com"
url = "https://example.com/pypi/simple"
default = false
secondary = false

# ...
```

From my development machine, I can authenticate with the [example.com](https://https://example.com/pypi/simple) repository and install `the-package`:

```sh
❯ poetry config http-basic.example.com $EXAMPLE_USER $EXAMPLE_PASSWORD

❯ poetry add --group dev the-package
Using version ^1.2.3 for the-package

Updating dependencies
Resolving dependencies... (6.7s)

Writing lock file

Package operations: 8 installs, 0 updates, 0 removals

  • Installing click-didyoumean (0.3.0)
  • Installing click-plugins (1.1.1)
  • Installing click-repl (0.2.0)
[ ... ]
```

I also expected to use the environment variables `POETRY_HTTP_BASIC_EXAMPLE_COM_USERNAME` and `POETRY_HTTP_BASIC_EXAMPLE_COM_PASSWORD` for installations during Docker image builds.

However, this does not work: [`poetry.config.config:230`](https://github.com/python-poetry/poetry/blob/dcd48c8df6d22246c21c0243fd387e3a9b189f93/src/poetry/config/config.py#L230) only replaces `-`s with `_`s, so it does not match these variables.

This request is to support other special characters in the repo name (e.g. `.` in `example.com`) instead of only `-`s.

**Comment by dimbleby** [comment]
you can simply give your repository a name that doesn't contain dots

of all the special characters that poetry might support here, dots are actually the least likely! Per line 219 they are used as a separator in the key names eg the configuration key that is actually being recovered here is `"http-basic.{name}.password"` - putting dots in `{name}` will break that.

I think this should be: thanks, but no thanks.

**Comment by bryant-finney** [comment]
@dimbleby 

> you can simply give your repository a name that doesn't contain dots

Yeah this is what I ended up doing since my intended convention managing repositories didn't work haha

> of all the special characters that poetry might support here, dots are actually the least likely! 

The `example.com` from the description (obviously) wasn't the actual convention I was trying to use. My repos were named by `group-name/repo-name` convention (I'm consolidating package storage locations and was hoping to match `poetry` repo names to GitLab repo names), and it took me a bit to figure out why it worked locally but failed in a GitLab pipeline 😄

> Per line 219 they are used as a separator in the key names eg the configuration key that is actually being recovered here is `"http-basic.{name}.password"` - putting dots in `{name}` will break that.

Ahh okay I think I follow: while 9b3440ee66171cb6fc8a327d52b948e16b3ac2f2 solves the issue I was having, it doesn't do anything for [config.py:235+](https://github.com/bryant-finney/poetry/blob/9b3440ee66171cb6fc8a327d52b948e16b3ac2f2/src/poetry/config/config.py#L235). Is that the error you're pointing out? (#7127 has the change I'm referring to)

Anywho, I appreciate the contribution so thanks anyways!

**Comment by neersighted** [comment]
> The example.com from the description (obviously) wasn't the actual convention I was trying to use. My repos were named by group-name/repo-name convention

This is why we ask for reproductions/issues to be as close to the actual problem/files you're trying to solve/use as possible, incidentally. We really strongly prefer people to provide as close to a copy-paste of what they are doing as corporately possible.

In this case, the dot is a no-go because dots have special meaning in Poetry config keys; other characters are... maybe problematic, but we already mostly support them, so I don't see a huge reason to say no. The key-parsing code should probably be a bit more robust/well-defined in the medium term, but again, that's evolutionary work that isn't made worse by what you propose.

**Comment by dimbleby** [comment]
the MR as it stands looks as though it allows dots in repository names, but _only_ if this configuration is accessed through an environment variable.

My feeling is that this isn't worth the trouble, I'd prefer to see an MR that clarifies the docs to say that repository should be given names made out of letters.  But that's, like, just my opinion...

**Comment by bryant-finney** [comment]
@neersighted 

> This is why we ask for reproductions/issues to be as close to the actual problem/files you're trying to solve/use as possible, incidentally. 

Sorry for any confusion! I misjudged what information was relevant here. But I'm kinda glad that I did, because I'm learning a lot that I would have missed otherwise haha

> In this case, the dot is a no-go because dots have special meaning in Poetry config keys

Yeah I had to poke around in a debug shell, but I completely see what you mean now. I tried the following and hit some issues towards the end:

```sh
# 1:
poetry source add some.repo https://gitlab.com/api/v4/projects/00000000/packages/pypi/simple/

# 2:
poetry config http-basic.some.repo <user> <pass>

# 3:
poetry add a-private-package
``` 

1. this command behaved as I would expect, and `pyproject.toml` was updated to include the following:
   ```toml
   # pyproject.toml
   [[tool.poetry.source]]
   name = "some.repo"
   url = "https://gitlab.com/api/v4/projects/00000000/packages/pypi/simple/"
   default = false
   secondary = false
   ```
2. this also appeared to work, but on closer look there was some nesting in `auth.toml` that I wasn't expecting:
   ```toml
   # auth.toml
   [http-basic]
   [http-basic.some]
   [http-basic.some.repo]
   username = "<user>"
   ```
3. this is where it broke: `poetry` was able to resolve the `a-private-package` version, so auth and connections with the private repo were functional at first. But when it reached the install operation, `poetry` received a `401` and failed to download the package.
   - in the debug shell; I saw that the [`Authenticator.request()`](https://github.com/python-poetry/poetry/blob/dcd48c8df6d22246c21c0243fd387e3a9b189f93/src/poetry/utils/authenticator.py#L257) was able to retrieve the credentials from the system keyring, and the first `GET` request passed. But the [`Executor._download_archive()`](https://github.com/python-poetry/poetry/blob/enhancement%2F7123%2Freplace-specials-with-underscores/src/poetry/installation/executor.py#L681) request hit an auth failure

@dimbleby I'm now thoroughly convinced that supporting periods in repo names is more trouble than it's worth: tracking down all of the locations impacted by the `.` sounds like a significant undertaking with little benefit.

I also agree that documenting this behavior would be helpful. I'm thinking a warning message emitted from `poetry source add` as well as a note in the [/docs/repositories](https://python-poetry.org/docs/repositories/) page would suffice; what do you think?

Also should I create a separate issue / PR for the docs update, or would updating the wording of this issue be fine?

**Comment by waweber** [comment]
I ran into a similar issue (#7156) but with repository names with underscores `_` in them. I'm not sure if that's also meant to be supported or if that should also be documented as problematic.
